extensions/amp-lightbox/1.0/lightbox.js

```diff
@@ -0,0 +1,190 @@
+ * @param {{current: (!LightboxDef.LightboxApi|null)}} ref
+ * @return {PreactDef.Renderable}
+ */
+function LightboxWithRef(
+  {
+    id,
+    animateIn = 'fade-in',
+    closeButtonAriaLabel,
+    children,
+    initialOpen,
+    onBeforeOpen,
+    onAfterClose,
+    enableAnimation,
+    ...rest
+  },
+  ref
+) {
+  // There are two phases to open and close.
+  // To open, we mount and render the contents (invisible), then animate the display (visible).
+  // To close, it's the reverse.
+  // `mounted` mounts the component. `visible` plays the animation.
+  const [mounted, setMounted] = useState(initialOpen);
+  const [visible, setVisible] = useState(false);
+  const classes = useStyles();
+  const lightboxRef = useRef();
+
+  // We are using refs here to refer to common strings, objects, and functions used.
+  // This is because they are needed within `useEffect` calls below (but are not depended for triggering)
+  const animateInRef = useRef(animateIn);
+  const onBeforeOpenRef = useRef(onBeforeOpen);
+  const onAfterCloseRef = useRef(onAfterClose);
+  const enableAnimationRef = useRef(enableAnimation);
+
+  useImperativeHandle(
+    ref,
+    () => ({
+      open: () => {
+        if (onBeforeOpenRef.current) {
+          onBeforeOpenRef.current();
+        }
+        setMounted(true);
+        setVisible(true);
+      },
+      close: () => {
+        setVisible(false);
+        setTimeout(() => setMounted(false), ANIMATION_DURATION);
+      },
+    }),
+    []
+  );
+
+  useEffect(() => {
+    const element = lightboxRef.current;
+    if (element == undefined) {
+      return;
+    }
+    let animation;
+
+    // "Make Visible" Animation
+    if (visible) {
+      const postVisibleAnim = () => {
+        setStyle(element, 'opacity', 1);
+        setStyle(element, 'visibility', 'visible');
+        element./*REVIEW*/ focus();
+      };
+      if (!element.animate || !enableAnimationRef.current) {
+        postVisibleAnim();
+        return;
+      }
+      animation = element.animate(ANIMATION_PRESETS[animateInRef.current], {
+        duration: ANIMATION_DURATION,
+        fill: 'both',
+        easing: 'ease-in',
+      });
+      animation.onfinish = postVisibleAnim;
+    } else {
+      // "Make Invisible" Animation
+      const postInvisibleAnim = () => {
+        setStyle(element, 'opacity', 0);
+        setStyle(element, 'visibility', 'hidden');
+        if (onAfterCloseRef.current) {
+          onAfterCloseRef.current();
+        }
+      };
+      if (!element.animate || !enableAnimationRef.current) {
+        postInvisibleAnim();
+        return;
+      }
+      animation = element.animate(ANIMATION_PRESETS[animateInRef.current], {
+        duration: ANIMATION_DURATION,
+        direction: 'reverse',
+        fill: 'both',
+        easing: 'ease-in',
+      });
+      animation.onfinish = postInvisibleAnim;
+    }
+    return () => {
+      if (animation) {
+        animation.cancel();
+        animation.onfinish();
```

Comment:
The problem is in this line. Here's the sequence that takes place when you go `visible -> !visible -> visible`:

1. `visible -> !visible`. Hiding animation is started. The `onfinish` will call `setMounted(false)`.
2. `!visible -> visible`. The cleanup function from (1) will be run. It will again run `onfinish`, and will again call `setMounted(false)`.

On the surface, the simple way to fix it is to reset the animation after it's done. E.g. 

```
const postInvisibleAnim = () => {
  setStyle(...);
  setMounted(false);
  animation = null;
};
```

That would make sense, since once animation is done, there's nothing to cancel anymore.

An alternative would be also to remove `animation.onfinish()` from the cleanup. Instead, we could add `animation.oncancel = postInvisibleAnim` or such. This option might be a bit cleaner. The one nit, however, it's best not to call `setMounted(false)` from `oncancel` I think. But I could be wrong on that.

PTAL and pick the one that fits the best with the current code and try it.

Reply:
First one looks good to me! 

Reasoning: Second one relies more on the Animation API, which is 1) still in a relatively new, developing state and more prone to change/errors and 2) not "future-proof" aka `oncancel` may be deprecated or changed in the future, leading us to hunting down API changes if made. I think doing less work is better in this case, where setting a animation to `null` will have 0 extraneous calls and lead us to the same result versus having 2 Animation API calls for canceling and finishing.

Reply:
Hmm. Thinking about this a bit more. I still don't see why we should call `onfinish()` here. I.e. if the "close" operation is being canceled, we really don't want to call `setMounted(false)` anymore...